drivers/spi/spi_wch.c

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2025 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#define DT_DRV_COMPAT wch_spi
+
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_wch);
+
+#include "spi_context.h"
+#include <errno.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/spi/rtio.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/math/ilog2.h>
+
+#include <ch32fun.h>
+
+#define SPI_CTLR1_LSBFIRST	((uint16_t)0x0080)
+#define SPI_CTLR1_BR_POS	3
+
+struct spi_wch_config {
+	SPI_TypeDef *regs;
+	const struct pinctrl_dev_config *pcfg;
+	const struct device *clk_dev;
+	uint8_t clk_id;
+};
+
+struct spi_wch_data {
+	struct spi_context ctx;
+};
+
+
+static uint8_t spi_wch_get_br(uint32_t target_clock_ratio)
+{
+	uint8_t div = 0;
+	int div_val = 2;
+
+	for (;div < 8; div++) {
+		if (div_val > target_clock_ratio) {
+			break;
+		}
+		div_val *= 2;
+	}
+	return div;
+}
+
+
+static int spi_wch_configure(const struct device *dev,
+			      const struct spi_config *config)
+{
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
```

Comment:
Add a `SPI_Typedef *regs = cfg->regs` and use it to replace the cfg->regs everywhere below.